### DISCUSSION TITLE
add a grouping to the view  or make it possible to include view

### DISCUSSION THREAD

**Discussion by viktorgusev1984** [likec4/likec4#533] [Ideas]
Hello. The tool is very good. An excellent replacement for the structurizr, but the structurizr has a deployment view.
There is a proposal to add a grouping to the view or make it possible to include views
example
```
view zone {

  group {
    title 'dc1'
    include element.tag=#dc1
  }
  group {
    title 'dc2'
    style {
     color secondary
    }
    include element.tag=#dc2
  }
}
```
mark tags components and then filter, it’s convenient
if you look to the side include view
```
view view1 {
  include *
}
view view2 {
  include *
}
view view2 {
  include view1
  include view2
}
```
it seems more difficult.
Connections between  group components are needed
no navigation required. groups for visual border only

**Comment by pavelpykhtin** [comment]
@davydkov , do you have any thoughts on how this should be implemented?
I see at least two options:
1. Use approach from structurizr: Add a separate 'deployment' layer to the DSL, which describes 'environments', 'instances' and 'deployment nodes'.
<details>
<summary>PlantUml</summary>

```
class Element
class Relationship
class ElementView


Element -* Element: part of
Element o- Relationship: connects
ElementView --o Relationship: includes
ElementView --o Element: includes


package Deployment{
    class DeploymentInstance
    class DeploymentView
    class DeploymentNode
    class Environment
}

DeploymentInstance o-- Element: instance of
DeploymentInstance --* DeploymentNode: hosted by
DeploymentView --o DeploymentInstance: includes
DeploymentView --o Relationship: includes
DeploymentNode o- Environment: includes
DeploymentView --o Environment: view of
```
</details>

2. Extend existing model with 'instance of' concept and allow including instances of elements into other elements.
<details>
<summary>PlantUML</summary>

```
class Element
class Relationship
class ElementView


Element -* Element: part of
Element o- Relationship: connects
ElementView --o Relationship: includes
ElementView --o Element: includes

class Instance
Instance --|> Element
Instance --* Element: part of
ElementView --o Instance: includes
```
</details>

I'd prefer the first option as it already works well in structurizr, allows to separate concepts of logical and physical architecture. But it bothers me that 'deployment view' and related concepts are part of C4 notation, while likec4 now is 'notation agnostic'.

The second option keeps the tool 'notation agnostic' and allows to show reusable components on diagram (which does not work well in C4 notation). But it will require predicates on each view to hide deployment aspects from logical architecture.

**Reply by lloydfischer** [reply]
@pavelpykhtin @davydkov  Depending on how one looks at this it is possible to see both of these options as the same thing - elements in the model that contain "instances" of elements that already existing in the model. As you know, but for the sake of those that don't, structurizr confines these instances (and nesting) to the special element, the deploymentNode. LikeC4, of course,  allows nesting everywhere. So an approach would be to also allow instance references anywhere and then LikeC4 could express the full structurizr model. 

As you state, then the views would have to be careful about what to expose. Further this "pollutes" the model with infrastucture, although the workspace support lets it be kept hidden. The top level infrastructure element can also be tagged as such and views could then easily filter this section out. 

A truly different approach (perhaps it could be called view driven) would be to allow the declaration of new elements and relationships and instance references to existing elements local to a view. This would localize the pollution while allowing the same level of expression. The downside, obviously, is that there is no longer a single model. 

I make no judgement of how feasible or difficult either of these approaches would be, and I'm sure there are more approaches. But in the time I have worked with structurizr these two approaches have seemed to me to be nice paths for enhancements. Since LikeC4 has already incorporated almost all my "nice ways to make structurizr better" these are ones I have left.

**Reply by davydkov** [reply]
Sorry for the late response.

> Use approach from structurizr: Add a separate 'deployment' layer to the DSL, which describes 'environments', 'instances' and 'deployment nodes'.

From my experience with Structurizr, the 'deployment' layer in DSL has always been 1-to-1 with the view, and I had the impression that the "deployment model" should directly correspond to the "deployment view". While I can see the benefits from an implementation perspective, I still have some doubts that this is the right approach.

> Extend existing model with 'instance of' concept and allow including instances of elements into other elements.

This could be an interesting idea, but it requires further investigation, as more questions have arisen. The straightforward idea would be to create an element type, like:

```zig
specification {
  element app
  element loggingComponent
}

model {
  app app1 {
    loggingComponent instance1
  }

  app app2 {
    loggingComponent instance2
  }
}
```

However, this doesn't work if I need each `loggingComponent` to have the same behavior, such as explicitly communicating with a logging service. I would have to define the same relationship for each instance.

Another edge case, as pointed out by Simon, can be found here: [Structurizr DSL Cookbook: Deployment Groups](https://docs.structurizr.com/dsl/cookbook/deployment-groups/).

At the moment, the most viable solution seems to be ad-hoc elements in the view, as @lloydfischer mentioned:

> allow the declaration of new elements and relationships, as well as instance references to existing elements, local to a view.

But I am not 100% sure in it

**Reply by davydkov** [reply]
@lloydfischer @pavelpykhtin 
Any thoughts for syntax?

**Reply by lloydfischer** [reply]
Very rough thoughts - I'll include them here simply to keep the conversation going. 

For including the instances in the model itself (not preferred) 
```
specification {
  element component
}
model {
  component system1 
  component system2 
  component system3 {
  instanceOf system1 id1 {
    metadata {
     ....
    }
  }
  instanceOf system2 id2 {
    metadata {
      ....
    }
  }
}
```

Using a view (preferred)
```
specification {
  element component
}
model {
  component system1 
  component system2 
}
views {
  view deployment {
      component system3 {
         instanceOf system1 id1 {
            metadata {
             ....
            }
          }
          instanceOf system2 id2 {
            metadata {
              ....
            }
          }
      }
  }
}
```

In the view case, an include statement would probably work as well as instanceOf to pull in an element from the model. 

The semantics I'm imagining are that anything done in an instanceOf stays there. Instances inherit from their template element. Relationships in the model would appear in the view following any normal view rules. Relationships added to the view itself would be local to that view (or extensions?). 

As an aside, many times the elements in the deployment view represent an elaboration of a relationship defined between components. The component model may show a relationship as Async, while in deployment we would introduce new components and relationships to represent brokers and other details. I mention this because it might be useful for to give a relationship an identity in the model and then link the elements and relationships in the deployment to the relationship in the model. Imagine drilling into a relationship like you would drill into a component.

**Comment by pavelpykhtin** [comment]
@davydkov , I don't like the idea of 1:1 relation between view and environment. Mostly because to me Architecture as Code is all about the model. Views are just a presentation of the model. Thus, views should not extend the model. But the amount of effort that we can put into development is quite limited. Therefore, implemented solution is better than "correct" but not implemented.

The second option suggested by @lloydfischer should work well for most of the cases.

**Reply by davydkov** [reply]
> Mostly because to me Architecture as Code is all about the model. Views are just a presentation of the model.

Yes, indeed, for me too.
But lets say we have this model:

```zig
specification {
  element element
}

model {
  element api
  element frontend
  frontend -> api
}

views {
  view index {
    title 'Landscape'
    include *
  }
}
```

this renders to
![index](https://github.com/user-attachments/assets/bba4c56d-5315-4827-9599-cb5f469f776c)

---

Now assume we have this:

```zig
specification {
  element element
  element zone
}

model {
  element api
  element frontend
  frontend -> api

  zone zone1 {
    api1 instanceof api
  }
  zone zone1 {
    api2 instanceof api
  }
}

views {
  view index {
    title 'Landscape'
    include *
  }
}
```

Without some additional rules how to interpret instanceOf and just consider them as part of the model,
for the same view we may result with:

![index](https://github.com/user-attachments/assets/7b22933b-bcda-4f9d-97e0-587db1cbd183)

**Comment by ilstephano** [comment]
May I add my 2cts here:
I'd prefer the first design - where the definition goes into the model.
I'm also not convinced of the 'deployment view' concept in structurizr and I don't see - in a modelling perspective - differences between 'logical' and 'physical' architectures.
I think reusing a model component - or let's state it better: defining a component once and using it multiple times - within the same model is natural, e.g. as class associations. 
So when it comes to me, a syntax like 
```
model {
  component system1 
  component system2 
  component system3 {
      system1 id1
      system2 id2
      component system31
      ...
  }
  component system4 {
      system1 id1
      system2 id2
      component system41
      ...
  }
}
```
would represent the 'grouping' concept, right?
Does that look reasonable? Or am I missing something?

**Reply by davydkov** [reply]
Looks good, but this is not possible, as for now syntax allows:

```zig
model {
  // Define element same name as kind
  // I use it often for `api`
  component = component "Component"
  
  system1 = component "System1" {
    component id1 // Ooops, what is it?
  }
  
}
```

We definitely need some additional syntax to express it.

**Reply by mtaylor** [reply]
"I think reusing a model component - or let's state it better: defining a component once and using it multiple times - within the same model is natural, e.g. as class associations."

Exactly this!

This feature would be a huge improvement.

**Comment by hubertmis** [comment]
Another idea I'm thinking about is separating the logical model (`model {}`) from the physical one (`deployment {}`?) with references between. Something like:

```zig
model {
  component service
  component storage {
    component interface
    component store
    component garbage_collector
    component backup

    interface -> store
    garbage_collector -> store
    backup -> store
  }

  service -> storage.interface 'Accesses data'
}

deployment {
  node main_cpu_cluster {
    instanceOf service
  }

  node active_database {
    instanceOf storage
  }

  mode backup_database {
    instanceOf storage
  }

  active_database.storage.backup -> backup_database.storage.backup 'Synchronizes data'
}

views {
  view landscape {
    include *; // renders logical model
  }

  deploymentView deployment {
    include *; // renders deployment at the landscape level
  }

  deploymentView active_database {
    include main_cpu_cluster
    include main_cpu_cluster.*
    include active_database
    include active_database.*
  }
}
```

**Reply by davydkov** [reply]
Wow, this is a really cool idea!

I had always felt that we needed to separate somehow physical elements from logical model.    
I consider a recipe with [multiple languages](https://langium.org/docs/recipes/multiple-languages/), where the physical model would be stored in files with another extension.  

But your approach seems both practical and neat. I think it’s worth exploring further. 

I can draft something up so we can investigate and develop this idea.

**Reply by Cyclonit** [reply]
That is a nice approach. If I am not mistaken, the `deployment` model will describe the physical (or virtualized) infrastructure that the logical services will be deployed on. I'd like to throw in some requirements/wishes as to what a system like that might support. 

1. **deployment / scaling groups**  
A logical service might be deployed multiple times across different zones/regions. Thus it'd be cool if we could visualize this.

2. **infrastructure / network partitions**  
Partitions like different networks or data centres are very important from an implementation perspective. The intersection points between them are a common source for problems (e.g. missing/wrongly configured firewall rules).

**Reply by Cyclonit** [reply]
Hi @davydkov,

which is the right place to start a more in depth discussion regarding deployment views? It is something I am very interested in, because it would allow likec4 to serve as a documentation tool bridging the gap between functional and technical documentation. It be nice if we good find a place to collect requirements and discuss possible solutions.

**Reply by BartDubois** [reply]
With this idea could it be possible to show the backup on the view?

I'm using version 1.53 it allow to make links with internal structure, like in the exemple, but do not allow to add to view 'active_database.storage.backup'. It report problem `Must reference deployment nodes or instances, but not internals`.

I will love to have such possibility.
